pkg/controller/backup_controller_test.go

```diff
@@ -422,3 +426,150 @@ func TestProcessBackup(t *testing.T) {
+			backup:          arktest.NewTestBackup().WithName("backup1").WithPhase(v1.BackupPhaseNew).WithVolumeSnapshotLocations(dupLocationNames),
+			locations:       arktest.NewTestVolumeSnapshotLocation().WithName(dupLocationNames[0]).WithProviderConfig(dupLocationList),
+			expectedSuccess: true,
+		},
+		{
+			name:            "multiple location names per provider",
+			backup:          arktest.NewTestBackup().WithName("backup1").WithPhase(v1.BackupPhaseNew).WithVolumeSnapshotLocations(multipleLocationNames),
+			locations:       arktest.NewTestVolumeSnapshotLocation().WithName(multipleLocationNames[0]).WithProviderConfig(multipleLocationList),
+			expectedSuccess: false,
+		},
+		{
+			name:             "no location name for the provider exists: the provider's default should be added",
+			backup:           arktest.NewTestBackup().WithName("backup1").WithPhase(v1.BackupPhaseNew),
+			defaultLocations: defaultLocationAWS,
+			expectedSuccess:  true,
+		},
+		{
+			name:            "no existing location name and no default location name given",
+			backup:          arktest.NewTestBackup().WithName("backup1").WithPhase(v1.BackupPhaseNew),
+			expectedSuccess: true,
+		},
+		{
+			name:             "multiple location names for a provider, default location name for another provider",
+			backup:           arktest.NewTestBackup().WithName("backup1").WithPhase(v1.BackupPhaseNew).WithVolumeSnapshotLocations(dupLocationNames),
+			locations:        arktest.NewTestVolumeSnapshotLocation().WithName(dupLocationNames[0]).WithProviderConfig(dupLocationList),
+			defaultLocations: defaultLocationGCP,
+			expectedSuccess:  true,
+		},
+	}
+
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			var (
+				client          = fake.NewSimpleClientset()
+				backupper       = &fakeBackupper{}
+				sharedInformers = informers.NewSharedInformerFactory(client, 0)
+				logger          = logging.DefaultLogger(logrus.DebugLevel)
+				pluginManager   = &pluginmocks.Manager{}
+				backupStore     = &persistencemocks.BackupStore{}
+			)
+			defer backupper.AssertExpectations(t)
+			defer pluginManager.AssertExpectations(t)
+			defer backupStore.AssertExpectations(t)
+
+			c := NewBackupController(
+				sharedInformers.Ark().V1().Backups(),
+				client.ArkV1(),
+				backupper,
+				false,
+				logger,
+				logrus.InfoLevel,
+				func(logrus.FieldLogger) plugin.Manager { return pluginManager },
+				NewBackupTracker(),
+				sharedInformers.Ark().V1().BackupStorageLocations(),
+				"default",
+				sharedInformers.Ark().V1().VolumeSnapshotLocations(),
+				test.defaultLocations,
+				metrics.NewServerMetrics(),
+			).(*backupController)
+
+			c.newBackupStore = func(*v1.BackupStorageLocation, persistence.ObjectStoreGetter, logrus.FieldLogger) (persistence.BackupStore, error) {
+				return backupStore, nil
+			}
+
+			// set up a Backup object to represent what we expect to be passed to backupper.Backup()
```

Comment:
L541-561 aren't really necessary - all you need to do is call `sharedInformers.Ark().V1().VolumeSnapshotLocations().Informer().GetStore().Add(...)` for each item in `test.locations`

Reply:
It seems I need to create a `v1.VolumeSnapshotLocatio` out of the test one. This is what I got if I used the test:

```
TestDefaultAndValidateSnapshotLocations/duplicate_locationName_per_provider:_should_filter_out_dups
panic: interface conversion: interface {} is *test.TestVolumeSnapshotLocation, not *v1.VolumeSnapshotLocation [recovered]
	panic: interface conversion: interface {} is *test.TestVolumeSnapshotLocation, not *v1.VolumeSnapshotLocation

```

Reply:
Ah. for each item in `test.location`, add `location.VolumeSnapshotLocation`.

Reply:
Ohhhh